### PR TITLE
Firmware switching

### PR DESCRIPTION
**Pull request by 13r0ck**
An improvement on the current production keyboard firmware flashing process. Rather than a separate script. Firmware flashing is part of the gui and uses .hex files rather than building from source. Done via a menu button in the header bar.

Notes:
- only in --launch-test mode
- can flash all 4 current models
- only possible to flash safe firmware:
    - heavy can flash to heavy and 2.0
    - 2.0 can flash to heavy and 2.0
    - 1.0 can only flash to 1.0
    - lite can only flash to lite
    
This is a feature request from @jackpot51

### CODE REVIEW

**Comment by ids1024** [comment]
I don't particularly like committing firmware `.hex` files to the repository, but the size isn't that large. Though this is also a license concern, since these are compiled from GPL source code. So there should probably be a link to the QMK repository with a clear indication of the commit these are built from.

Perhaps the firmware could be downloaded by the Configurator (from LVFS?), but that's a bit more complicated to do.

**Comment by 13r0ck** [comment]
@jackpot51 suggested adding the hex files to the repo

File: backend/src/board.rs [hunk 1]
```diff
@@ -230,6 +231,21 @@ impl Board {
...
+        static RE: Lazy<Regex> = Lazy::new(|| Regex::new("system76/launch_lite_.*").unwrap());
+        RE.is_match(self.model())
+    }
+
+    pub fn is_1(&self) -> bool {
+        static RE: Lazy<Regex> = Lazy::new(|| Regex::new("system76/launch_1.*").unwrap());
+        RE.is_match(self.model())
+    }
+
+    pub fn is_2(&self) -> bool {
```
Comment (ids1024) [inline backend/src/board.rs:244]: Why are `launch_heavy_1` and `launch_2` being grouped together here?

Comment (13r0ck) [reply backend/src/board.rs:244]: Because they do the same thing. Production needs to be able to flash heavys as 2s, and 2s as heavys.

Comment (13r0ck) [reply backend/src/board.rs:244]: They both use the same launch 2.0 PCB

Comment (ids1024) [reply backend/src/board.rs:244]: Makes sense. Might be worth having a comment here mentioning that.

**Review by leviport — DISMISSED** [review]

**Review by leviport — CHANGES_REQUESTED** [review]
I think there might be a problem with this. Both my Lite and Launch 1 end up showing me a "Firmware update required!" message, but the extra menu button doesn't have any options in it with which to actually update them. Both are running current `master` firmware from our QMK fork, in case that's related.

**Comment by 13r0ck** [comment]
ya that is just a mismatch of the version numbering of the .hex files, and what fwupd has. Not a deal breaker as production could still fix that by updating the keyboard firmware in gnome-control-center (this is what will happen if we update fwupd in the future, but forget to update the hex files). Though that should be fixed with 8d5ed18

**Comment by 13r0ck** [comment]
I rebased this on #173  to fix the build

**Comment by 13r0ck** [comment]
Thank you. This is now rebased on master, and ready again for QA

**Review by leviport — CHANGES_REQUESTED** [review]
I'm now testing on a different machine since fwupdmgr is a mess on my work machine, but now I'm running into something strange. When I plug in a Heavy, it is detected and shown (has current firmware, of course), and I see the option to flash to launch_2 or launch_heavy_1. However when I click either option, I get `dfu-programmer: no device present.`

So then I disconnect the Heavy and reconnect it with Esc held to put it into bootloader mode, thinking `avrdude` is waiting, but nothing happens. 

Then if I click one of the flash buttons in the menu again, I see this output in the terminal that I was running the configurator from:
```
Error opening the file.
Something went wrong with creating the memory image.
```

So that does appear to have a bug preventing me from flashing, but I'm also a bit uneasy with the lack of visual indications in the configurator window telling me how to flash the board, or what is happening. I know this is just an internal tool, but if we can make it easier on Assembly, then I'm all for it.

Also, the buttons don't change from the last detected board (I mean, it's probably not possible to do so anyway if only an AVR bootloader is detected). This means that I can plug in a Heavy, then unplug it, then plug in a Lite with Esc held, and still be shown a "flash launch_heavy_1" button. This would put the Lite in a state that would require the external programmer to recover from (I've done it before :sweat_smile: ). I'm not sure if that's just something to be careful of, or if there's a way we could make sure we couldn't kill a Lite with this method, even if it would have to be done somewhat deliberately.

**Comment by ids1024** [comment]
> Also, the buttons don't change from the last detected board

Good catch, I didn't notice that looking at the code. This is important is assembly if flashing multiple kinds of boards from the same computer.

These should be `kbd.` actions in `src/keyboard.rs`, like `invert-f-keys`.

**Comment by 13r0ck** [comment]
That is intentional. This doesn't need to be perfect. Production needs to ship keyboards.

**Comment by 13r0ck** [comment]
> This is important is assembly

Production does the keyboard flashing. They are different teams. Both deserving of at least enough respect to be called by the right name.

**Comment by leviport** [comment]
>Avrdude doesn't wait. This doesn't use the qmk make files. avrdude never waits. It is the qmk make files that do

Fair point, but I've still not been able to successfully flash something. If there is a procedure I'm unaware of that will be circulated internally, it would be very helpful for me to use during testing. I called it a bug earlier, but it also kinda feels like I'm just doing it wrong.

**Comment by 13r0ck** [comment]
No worries. Work from home is driving me insane. I'm coming in tomorrow in hazmat if I have to, we can hash this all out then. Have a good night!

**Comment by leviport** [comment]
Thanks, you too!

**Comment by 13r0ck** [comment]
This now has a lot more improved UX:
1. The flash menu only appears when a keyboard is plugged in in bootloader mode
    - this prevents the edge case of plugging in a keyboard, then plugging in a bootloader mode keyboard of another model
    - this also very clearly shows when the keyboard has connected in booloader mode, so no need to click the flash button blindly hoping it finds the keyboard
2. This starts/resets the keyboard once done, so it is clear that the flash is done
    - big lights when workie
    - the keyboard is then auto-detected in the configurator, so testing can begin without needing to (un/re)plug un the keyboard
3. The firmware version for lite/launch_1 is now actually correct.

**Review by leviport — APPROVED** [review]
Works very nicely now! 

Also partially addresses https://github.com/pop-os/keyboard-configurator/issues/153 by deactivating the USB section when a Lite is plugged in.

**Review by ids1024 — APPROVED** [review]
Not the best way to deal with the `GAction`s, but it seems fine if this meets production's requirements.

The backend changes will conflict with https://github.com/pop-os/keyboard-configurator/pull/169, which is no longer using glib there, but that shouldn't be too hard to resolve after this is merged.

**Comment by 13r0ck** [comment]
> Not the best way to deal with the `GAction`

Mind elaborating please?

**Comment by ids1024** [comment]
As I mentioned earlier the actions should be provided by the `Keyboard`, like `invert-f-keys`. Though I actually not sure if there a "correct" way to completely hid actions from the menu, or set their names dynamically. Normally they are just made insensitive.

But it doesn't make much difference, especially if we're going to eventually port the UI to Iced.

**Pull request merged**